### PR TITLE
logic and rspec for update book request

### PR DESCRIPTION
**Pull request by banguyenvareal**

### CODE REVIEW

File: app/controllers/users/book_requests_controller.rb [hunk 1]
```diff
@@ -8,13 +11,67 @@ def index
...
+        redirect_to user_book_request_path(current_user, @book_request)
+      else
+        render 'edit'
+      end
     end
 
     private
 
     def book_requests_user
       User.find(params[:user_id])
```
Comment (mrcuong84) [inline app/controllers/users/book_requests_controller.rb]: find_by?

File: spec/controllers/users/book_request/update_spec.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+          user_id: current_user.id, id: book_request.id,
+          book_request: invalid_book_request_params
+        }
+      end
+      it 'number of book request not change' do
+        subject
+        expect { subject }.to change(BookRequest, :count).by(0)
+      end
+
+      it 'should be reder to edit' do
```
Comment (mrcuong84) [inline spec/controllers/users/book_request/update_spec.rb]: reder --> render

File: app/controllers/users/book_requests_controller.rb [hunk 3]
```diff
@@ -8,13 +11,67 @@ def index
...
+
+    def update
+      @book_request = BookRequest.find(params[:id])
+      if @book_request.update_attributes(book_request_params)
+        destroy_image_files unless image_file_params
+        create_image_files
+        flash[:success] = t('book_requests.update.success')
+        redirect_to user_book_request_path(current_user, @book_request)
+      else
+        render 'edit'
```
Comment (mrcuong84) [inline app/controllers/users/book_requests_controller.rb:26]: should add warning flash

Comment (mrcuong84) [reply app/controllers/users/book_requests_controller.rb:26]: and then add in rspec for this case

File: spec/controllers/users/book_request/update_spec.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        expect(subject).to render_template(:edit)
+      end
+    end
+
+    context 'with valid params' do
```
Comment (mrcuong84) [inline spec/controllers/users/book_request/update_spec.rb]: check destroy images success?
and then create images success of not ?

File: app/controllers/users/book_requests_controller.rb (hunk 3 above)
Comment (banguyenvareal) [reply app/controllers/users/book_requests_controller.rb:26]: warning ready in edit template(from shared/error_message)

File: app/controllers/users/book_requests_controller.rb [hunk 5]
```diff
@@ -8,13 +11,66 @@ def index
...
+    end
+
+    def create_image_files
+      image_file_params&.each do |image_file|
+        @book_request.book_request_images.create(file: image_file)
+      end
+    end
+
+    def destroy_image_files
+      file_images = @book_request.book_request_images
```
Comment (dinhvareal) [inline app/controllers/users/book_requests_controller.rb:79]: remove it

File: app/views/books/new.html.erb [hunk 6]
```diff
@@ -1,3 +1,3 @@
+<h1 class="horizontal-border"><%= t('book.creat_label') %></h1>
+<%= render 'books/form', submit_label: t('book.create') %>
 <h1 class="horizontal-border"><%= t('book.create_label') %></h1>
```
Comment (dinhvareal) [inline app/views/books/new.html.erb]: same content with l1

File: config/locales/en.yml [hunk 7]
```diff
@@ -150,3 +150,5 @@ en:
   book_requests:
     create:
       success: "A new book request created"
+    update:
+      success: "your book request updated"
```
Comment (dinhvareal) [inline config/locales/en.yml]: `your` -> `Your`

File: spec/controllers/users/book_request/edit_spec.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+      it 'redirect to root page' do
+        expect(subject).to redirect_to(root_url)
+      end
+    end
+
+    context 'book request not belong to current user' do
+      include_context 'logged in'
+      before do
+        get :edit, params: { user_id: current_user.id, id: random_id }
```
Comment (dinhvareal) [inline spec/controllers/users/book_request/edit_spec.rb]: why is `random_id`?

File: spec/controllers/users/book_request/edit_spec.rb [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        expect(flash.count).to equal(1)
+        expect(flash[:danger]).to eql(I18n.t('not_found'))
+      end
+
+      it 'redirect to root page' do
+        expect(subject).to redirect_to(root_url)
+      end
+    end
+
+    context 'with valid params' do
```
Comment (dinhvareal) [inline spec/controllers/users/book_request/edit_spec.rb]: rename this context

File: spec/controllers/users/book_request/update_spec.rb [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        put :update, params: {
+          user_id: current_user.id, id: book_request.id,
+          book_request: invalid_book_request_params
+        }
+      end
+      it 'book request not change' do
+        expect { subject }.to_not change(book_request, :reload)
+      end
+
+      it 'number of book request not change' do
```
Comment (dinhvareal) [inline spec/controllers/users/book_request/update_spec.rb]: similar to line 16

**Review by dinhvareal — APPROVED** [review]

File: app/controllers/users/book_requests_controller.rb [hunk 11]
```diff
@@ -1,5 +1,8 @@
 module Users
   class BookRequestsController < ApplicationController
+    before_action :check_log_in, only: %i[new edit index update]
+    before_action :check_valid_user, only: %i[edit update]
+    before_action :find_and_assign_book_request, only: %i[edit update]
     def index
       @book_requests = book_requests_user.book_requests
```
Comment (hieunguyentrung) [inline app/controllers/users/book_requests_controller.rb:7]: `book_requests_user == nil` thì sao ?

File: app/controllers/users/book_requests_controller.rb [hunk 12]
```diff
@@ -8,13 +11,66 @@ def index
...
     end
 
     private
 
     def book_requests_user
-      User.find(params[:user_id])
+      User.find_by(params[:user_id])
+    end
+
+    def handle_danger_flash
```
Comment (hieunguyentrung) [inline app/controllers/users/book_requests_controller.rb]: `set_not_found_flash_and_redirect`

File: app/controllers/users/book_requests_controller.rb [hunk 13]
```diff
@@ -8,13 +11,66 @@ def index
...
+      flash[:danger] = t('not_found')
+      redirect_to root_url
```
Comment (hieunguyentrung) [inline app/controllers/users/book_requests_controller.rb]: `redirect_to(root_url) && return`

File: app/controllers/users/book_requests_controller.rb [hunk 14]
```diff
@@ -8,13 +11,66 @@ def index
...
+    end
+
+    def check_valid_user
```
Comment (hieunguyentrung) [inline app/controllers/users/book_requests_controller.rb]: `handle_invalid_user!`

File: app/controllers/users/book_requests_controller.rb [hunk 15]
```diff
@@ -8,13 +11,66 @@ def index
...
+      handle_danger_flash unless current_user.id == params[:user_id].to_i
+    end
+
+    def check_log_in
```
Comment (hieunguyentrung) [inline app/controllers/users/book_requests_controller.rb]: `authenticate_user!`

File: app/controllers/users/book_requests_controller.rb [hunk 16]
```diff
@@ -8,13 +11,66 @@ def index
...
+      @book_request = BookRequest.find_by(id: params[:id])
+    end
+
+    def book_request_params
+      params.require(:book_request).permit(
+        :name, :comment, :budget, :quantity
+      )
+    end
+
+    def image_file_params
```
Comment (hieunguyentrung) [inline app/controllers/users/book_requests_controller.rb]: `image_files_params`

File: app/controllers/users/book_requests_controller.rb [hunk 17]
```diff
@@ -8,13 +11,66 @@ def index
...
+
+    def create_image_files
+      image_file_params&.each do |image_file|
+        @book_request.book_request_images.create(file: image_file)
+      end
+    end
+
+    def destroy_image_files
+      file_images = @book_request.book_request_images
+      file_images&.each do |image_file|
```
Comment (hieunguyentrung) [inline app/controllers/users/book_requests_controller.rb]: `file_images.destroy_all if file_images.present?`

**Review by hieunguyentrung — CHANGES_REQUESTED** [review]

**Review by dinhvareal — APPROVED** [review]

**Review by mrcuong84 — APPROVED** [review]

**Review by hieunguyentrung — APPROVED** [review]

**Pull request merged**